### PR TITLE
update language detection model to v0.13.9

### PR DESCRIPTION
**Pull request by sarnoult**
This updates the language detection model in text to v0.13.9.

Besides:
- The model components are renamed from scanner/encoder to encoder/classifier
- Test accuracy is added for evaluation

### CODE REVIEW

File: text/lang-detection/Project.toml [hunk 1]
```diff
@@ -4,10 +4,4 @@ Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
 Gumbo = "708ec375-b3d6-5a57-a7ce-8257bf98657a"
 HTTP = "cd3eb016-35fb-5094-929b-558a96fad6f3"
 Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
-Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
-Unicode = "4ec0a83e-493e-50e2-b9ac-8f72acf5a8f5"
-
-[compat]
-Flux = "0.11.4"
-julia = "1.5"
```
Comment (ToucheSir) [inline text/lang-detection/Project.toml:13]: These deps and compat entries should be kept. The latter just need a bump, and if you wouldn't mind I think adding compat entries for the other packages would be good too.

Comment (ToucheSir) [reply text/lang-detection/Project.toml:13]: That said, removing Parameters in favour of `Base.@kwdef` could save one dependency if you're up for it.

Comment (sarnoult) [reply text/lang-detection/Project.toml]: Thank you. I have changed the parameters to `Base.@kwdef` and updated `Project.toml`. For the compat entries, I tested with julia 1.6.6; other packages than Flux are just the latest.

File: text/lang-detection/Project.toml [hunk 2]
```diff
@@ -3,11 +3,13 @@ Cascadia = "54eefc05-d75b-58de-a785-1a3403f0919f"
...
 Unicode = "4ec0a83e-493e-50e2-b9ac-8f72acf5a8f5"
 
 [compat]
-Flux = "0.11.4"
-julia = "1.5"
+Cascadia = "1.0.2"
+Flux = "0.13.9"
+Gumbo = "0.8.2"
+HTTP = "1.7.4"
+julia = "1.6"
```
Comment (ToucheSir) [inline text/lang-detection/Project.toml:15]:
```suggestion
Cascadia = "1"
Flux = "0.13.9"
Gumbo = "0.8"
HTTP = "1.7"
julia = "1.6"

```
Just simplifying compat a bit and adding back in the final newline.

File: text/lang-detection/model.jl [hunk 3]
```diff
@@ -61,67 +61,79 @@ function get_processed_data(args)
...
     @info("Constructing Model...")
-    scanner, encoder = build_model(args)
+    model = build_model(args)
+    loss(model, x, y) = logitcrossentropy(model(x), y)
+    opt = Flux.setup(ADAM(args.lr), model)
+
+    @info("Training...") 
+    for epoch in 1:args.epochs
+        Flux.train!(loss, model, trainData, opt)
+        test_loss = sum(loss(model, x, y) / length(testX) for (x, y) in zip(testX, testY))
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:130]:
```suggestion
        test_loss = mean(loss(model, x, y) for (x, y) in zip(testX, testY))
```

File: text/lang-detection/model.jl [hunk 4]
```diff
@@ -61,67 +61,79 @@ function get_processed_data(args)
...
 
     train, test = dataset[1:end-args.test_len], dataset[end-args.test_len+1:end]
-    return train, test
+    (testX, testY) = unzip(test)
+    return train, testX, testY, langs
 end
 
+# https://stackoverflow.com/questions/36367482/unzip-an-array-of-tuples-in-julia
+unzip(a) = (getfield.(a, x) for x in fieldnames(eltype(a)))
+
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:75]:
```suggestion
```
Removed in favour of suggestion in next comment (on `(testX, testY) = unzip(test)` line).

File: text/lang-detection/model.jl [hunk 5]
```diff
@@ -61,67 +61,79 @@ function get_processed_data(args)
...
     ## See which chars will be represented as "unknown"
-    unique(filter(x -> x ∉ alphabet, join(vcat(values(corpora)...))))
-
+    unkChars = unique(filter(x -> x ∉ alphabet, join(vcat(values(corpora)...))))
+    
     dataset = [(onehotbatch(s, alphabet, '_'), onehot(l, langs)) for l in langs for s in corpora[l]] |> shuffle
 
     train, test = dataset[1:end-args.test_len], dataset[end-args.test_len+1:end]
-    return train, test
+    (testX, testY) = unzip(test)
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:69]:
```suggestion
    testX, testY = first.(test), last.(test)
```
I leave it to your discretion whether a comment is necessary :)

**Review by ToucheSir — COMMENTED** [review]
Thanks for updating. A couple small changes and I think we're good to go.

File: text/lang-detection/model.jl (hunk 5 above)
Comment (sarnoult) [reply text/lang-detection/model.jl]: Excellent, thank you! I have incorporated your comments.

File: text/lang-detection/model.jl [hunk 6]
```diff
@@ -61,67 +61,76 @@ function get_processed_data(args)
     args.alphabet_len = length(alphabet)
 
     ## See which chars will be represented as "unknown"
-    unique(filter(x -> x ∉ alphabet, join(vcat(values(corpora)...))))
-
+    unkChars = unique(filter(x -> x ∉ alphabet, join(vcat(values(corpora)...))))
+
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:65]:
```suggestion
    unk_chars = unique(filter(∉(alphabet), join(vcat(values(corpora)...))))
```

File: text/lang-detection/model.jl [hunk 7]
```diff
@@ -61,67 +61,76 @@ function get_processed_data(args)
...
-
+# and one [LSTM](https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.LSTM) layers, and encodes it through 
+# the state of its last character.
+# The **classifier** inputs this encoding and outputs the predicted language for the sentence.
+# The model is defined as a [Custom model](https://fluxml.ai/Flux.jl/stable/models/advanced/)
+
+struct EncoderClassifier
+    encoder
+    classifier
+end
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:85]:
```suggestion
struct EncoderClassifier{E, C}
    encoder::E
    classifier::C
end
```

File: text/lang-detection/model.jl [hunk 8]
```diff
@@ -61,67 +61,76 @@ function get_processed_data(args)
...
 # Notice that we use the function [reset!](https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.reset!) 
 # when computing the model's prediction to reset the hidden state of an LSTM layer back to its original value.
 
-
-function model(x, scanner, encoder)
-    state = scanner.(x.data)[end]
-    reset!(scanner)
-    encoder(state)
+function (m::EncoderClassifier)(x)
+    state = m.encoder(x)[:,end]
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:97]:
```suggestion
    state = m.encoder(x)[:, end]
```

File: text/lang-detection/model.jl [hunk 9]
```diff
@@ -61,67 +61,76 @@ function get_processed_data(args)
...
-# Also, it creates a callback function to output the test loss as defined in the args object.   
-
 
 function train(; kws...)
     ## Initialize Hyperparameters
     args = Args(; kws...)
     
     ## Load Data
-    train_data, test_data = get_processed_data(args)
+    trainData, testX, testY, langs = get_processed_data(args)
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:117]:
```suggestion
    train_data, test_X, test_Y, langs = get_processed_data(args)
```

File: text/lang-detection/model.jl [hunk 10]
```diff
@@ -61,67 +61,76 @@ function get_processed_data(args)
...
-    testloss() = mean(loss(t...) for t in test_data)
-    
-    opt = ADAM(args.lr)
-    ps = params(scanner, encoder)
-    evalcb = () -> @show testloss()
-    @info("Training...")
-    Flux.train!(loss, ps, train_data, opt, cb = throttle(evalcb, args.throttle))
+    testPredictions = [onecold(model(x), langs) for x in testX]
+    accuracy = mean(testPredictions .== [onecold(y, langs) for y in testY])
+    @show accuracy
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:133]:
```suggestion
    test_predictions = [onecold(model(x), langs) for x in test_X]
    accuracy = mean(test_predictions .== [onecold(y, langs) for y in test_Y])
    @show accuracy
```

File: text/lang-detection/model.jl [hunk 11]
```diff
@@ -61,67 +61,76 @@ function get_processed_data(args)
...
+    model = build_model(args)
+    loss(model, x, y) = logitcrossentropy(model(x), y)
+    opt = Flux.setup(ADAM(args.lr), model)
+
+    @info("Training...") 
+    for epoch in 1:args.epochs
+        Flux.train!(loss, model, trainData, opt)
+        test_loss = mean(loss(model, x, y) for (x, y) in zip(testX, testY))
+        @show epoch, test_loss
+    end
```
Comment (ToucheSir) [inline text/lang-detection/model.jl:129]:
```suggestion
    @info("Training...") 
    for epoch in 1:args.epochs
        Flux.train!(loss, model, train_data, opt)
        test_loss = mean(loss(model, x, y) for (x, y) in zip(test_X, test_Y))
        @show epoch, test_loss
    end
```
